test/gtest/core/test_mc_reduce.cc

```diff
@@ -0,0 +1,274 @@
+            res_d = (typename T::type *)res_d_mc_header->addr;
+            ucc_mc_alloc(&buf1_d_mc_header, n_bytes, mtype);
+            buf1_d = (typename T::type *)buf1_d_mc_header->addr;
+            ucc_mc_alloc(&buf2_d_mc_header, n * n_bytes, mtype);
+            buf2_d = (typename T::type *)buf2_d_mc_header->addr;
+            ucc_mc_memcpy(res_d, res_h, n_bytes, mtype, UCC_MEMORY_TYPE_HOST);
+            ucc_mc_memcpy(buf1_d, buf1_h, n_bytes, mtype, UCC_MEMORY_TYPE_HOST);
+            ucc_mc_memcpy(buf2_d, buf2_h, n * n_bytes, mtype,
+                          UCC_MEMORY_TYPE_HOST);
+            buf1 = buf1_d;
+            buf2 = buf2_d;
+            res  = res_d;
+        } else {
+            buf1 = buf1_h;
+            buf2 = buf2_h;
+            res  = res_h;
+        }
+
+        return UCC_OK;
+    }
+
+    ucc_status_t free_bufs(ucc_memory_type_t mtype)
+    {
+        if (buf1_h != nullptr) {
+            ucc_mc_free(buf1_h_mc_header);
+        }
+        if (buf2_h != nullptr) {
+            ucc_mc_free(buf2_h_mc_header);
+        }
+        if (res_h != nullptr) {
+            ucc_mc_free(res_h_mc_header);
+        }
+        if (buf1_d != nullptr) {
+            ucc_mc_free(buf1_d_mc_header);
+        }
+        if (buf2_d != nullptr) {
+            ucc_mc_free(buf2_d_mc_header);
+        }
+        if (res_d != nullptr) {
+            ucc_mc_free(res_d_mc_header);
+        }
+
+        return UCC_OK;
+    }
+
+    virtual void TearDown() override
+    {
+        free_bufs(mem_type);
+        ucc_mc_finalize();
+    }
+
+    void test_reduce(ucc_memory_type_t mt) {
+        ucc_status_t status;
+
+        if (UCC_OK !=  ucc_mc_available(mt)) {
+            GTEST_SKIP();
+        }
+        ASSERT_EQ(this->alloc_bufs(mt, 1), UCC_OK);
+        status = ucc_mc_reduce(this->buf1, this->buf2,
+                               this->res, this->COUNT, T::dt, T::redop, mt);
+        if (UCC_ERR_NOT_SUPPORTED == status) {
+            GTEST_SKIP();
+        }
+        ASSERT_EQ(status, UCC_OK);
+
+        if (mt != UCC_MEMORY_TYPE_HOST) {
+            ucc_mc_memcpy(this->res_h, this->res_d, this->COUNT * sizeof(*this->res_d),
+                          UCC_MEMORY_TYPE_HOST, mt);
+        }
+        for (int i = 0; i < this->COUNT; i++) {
+            T::assert_equal(T::do_op(this->buf1_h[i],
+                                     this->buf2_h[i]), this->res_h[i]);
+        }
+    };
+
+    void test_reduce_multi(ucc_memory_type_t mt) {
+        const int    num_vec = 3;
+        ucc_status_t status;
+
+        if (UCC_OK !=  ucc_mc_available(mt)) {
+            GTEST_SKIP();
+        }
+        ASSERT_EQ(this->alloc_bufs(mt, num_vec), UCC_OK);
+        status = ucc_mc_reduce_multi(this->buf1, this->buf2, this->res,
+                       num_vec, this->COUNT, this->COUNT*sizeof(*this->buf2),
+                       T::dt, T::redop, mt);
+        if (UCC_ERR_NOT_SUPPORTED == status) {
```

Comment:
we make pairs so that reduction is always supported then this check is not needed

Reply:
For test_mc_reduce (not collective but just reduction backend) i don't enumerate separate dtypes for CUDA. I have the enumerations for ALL possible pairs, but skip not-supported. I could create separate set of testing::types for CUDA only. It will be little bit more explicit but then i can remove test skip here. What do you prefer?

Reply:
ok, keep it as is pls, i thought we support all ops and dtypes in cuda